gcsio/src/main/java/com/google/cloud/hadoop/gcsio/ZeroCopyMessageMarshaller.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2021 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.hadoop.gcsio;
+
+import com.google.protobuf.ByteString;
+import com.google.protobuf.CodedInputStream;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.MessageLite;
+import com.google.protobuf.Parser;
+import com.google.protobuf.UnsafeByteOperations;
+import io.grpc.Detachable;
+import io.grpc.HasByteBuffer;
+import io.grpc.KnownLength;
+import io.grpc.MethodDescriptor.PrototypeMarshaller;
+import io.grpc.protobuf.lite.ProtoLiteUtils;
+import io.grpc.Status;
+import io.grpc.MethodDescriptor;
+import java.io.InputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+
+// Custom gRPC marshaller to use zero memory copy feature of gRPC when deserializing messages.
+// This achieves zero-copy by deserializing proto messages pointing to the buffers in the input
+// stream to avoid memory copy so stream should live as long as the message can be referenced.
+// Hence, it exposes the input stream to applications (through popStream) and applications are
+// responsible to close it when it's no longer needed. Otherwise, it'd cause memory leak.
+class ZeroCopyMessageMarshaller<T extends MessageLite> implements PrototypeMarshaller<T> {
+  private Map<T, InputStream> unclosedStreams = Collections.synchronizedMap(new IdentityHashMap<>());
+  private final Parser<T> parser;
+  private final PrototypeMarshaller<T> baseMarshaller;
+
+  ZeroCopyMessageMarshaller(T defaultInstance) {
```

Comment:
how about we rename `defaultInstance` to `message` Or `messageInstance`?

Reply:
This is to be consistent with the name of variable of gRPC-Java.
https://github.com/grpc/grpc-java/blob/f2ed41aca0504c283f943d9468ef316906d1f117/protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java#L82